File: src/karkinolution/tick/worldcycle.py
from karkinolution.organism.identity import EntityTypes

from karkinolution.terrain.world import World

from karkinolution.tick.lifecycle import (
    RunnerCorpse,
    RunnerCreature,
)


class Runner:
    @staticmethod
    def run(world:World):
        territory = world.territory
        entities = world.entities
        entity_map = world.entity_map

        ### RUN CELLS
        for cell in territory.values:
            cell.pass_time()


        ### RUN CREATURES
        for coord, id in list(entity_map.iter):
            cell = territory.get(coord)
            if id.e_type == EntityTypes.CREATURE:
                creature = entities.get_creature(id)

                RunnerCreature.run_creature(creature, world)
            elif id.e_type == EntityTypes.CORPSE:
                corpse = entities.get_corpse(id)

                RunnerCorpse.run_corpse(corpse, entity_map, entities)
        world.time+=1

File: src/karkinolution/ticks/worldcycle.cpp
#include <karkinolution/ticks/worldcycle.hpp>
#include <karkinolution/organism/ontology.hpp>
#include <karkinolution/world/world.hpp>
#include <karkinolution/ticks/lifecycle.hpp>
#include <karkinolution/terrain/soil.hpp>

void RunnerWorld::run(World& world) {
    for (auto& soil : world.territory.values()) {
        if (soil.components.exists<FoodState>()) {
            auto* ptr = soil.components.try_get<FoodState>();
            ptr->food += ptr->regen_tax;
        }
    }

    std::vector<Id> entities_to_run;

    for (const auto& [coord, id] : world.entity_map.iter()) {
        entities_to_run.push_back(id);
    }

    for (const auto& id : entities_to_run) {
        if (id.entity_type == EntityTypes::CREATURE) {
            RunnerCreature::run_creature(
                world.entities.at_creature(id),
                world
            );
        } 
        else if (id.entity_type == EntityTypes::CORPSE) {
            RunnerCorpse::run_corpse(
                world.entities.at_corpse(id),
                world
            );
        }
    }

    world.time += 1;
}